libimage/manifest_list.go

```diff
@@ -26,6 +26,12 @@ import (
 // NOTE: the abstractions and APIs here are a first step to further merge
 // `libimage/manifests` into `libimage`.
 
+// Copier is a simple interface declaration for the existing copier
+type Copier interface {
+	Copy(ctx context.Context, source, destination types.ImageReference) ([]byte, error)
+	Close() error
+}
```

Comment:
As a general matter of Go API design, I now think libraries exporting objects should _not_ be publishing interface types if at all possible.

(Sure, export an interface type if the caller needs to provide an object, to define what the object is expected to satisfy.)

The reason is that adding a method to an interface breaks all other implementations, so there is really no way to have an API-stable exported interface for an evolving feature set; the only API-stable way to use a such a Go interface which expands over time is to only ever have that one implementation — and at that point there doesn’t need to be an interface at all.

Instead:
- For the purpose of unit testing / mocking: define a private interface in the test subpackage of the _consumer_ of the object
- For the purpose of limiting the method set of the exported type (e.g. if the implementation struct must have public methods to conform to some other interface): instead, export a wrapper struct that _only_ has the truly-public methods, and make the actual implementation a private structure not visible to external callers. (An extreme variant of this is going one step further in https://github.com/containers/image/blob/main/internal/signer/signer.go , where the only publicly-visible method is `Close`; and all other methods can be called internally from c/image. That’s only necessary if the module-private methods need to be called from multiple subpackages.

---

Also, if `NewCopier` is public, the returned type probably should also be public, to allow callers to e.g. store the returned value in a struct field; if the returned type is unnamed, it can _only_ be used via private `:=` variables — or using an interface type.